### PR TITLE
delete nodegroups with config file

### LINKED ISSUE #664: delete missing nodegroups based on config file
This would improve upgrade docs.

### PR DESCRIPTION
**Pull request by errordeveloper**
### Description

Enhance `eksctl delete nodegroup`:

- add functionality to delete multiple nodegroups with `--config-file`
- and `--include`/`--exclude` filters
- add `--only-missing` (inverted fitering)
- add `--approve` (only with `--config-file`)

Fixes #664.

### Checklist

- [x] Code compiles correctly (i.e `make build`)
- [x] Added tests that cover your change (if possible)
- [x] All unit tests passing (i.e. `make test`)
- [x] All integration tests passing (i.e. `make integration-test`)
- [x] Added/modified documentation as required (such as the `README.md`, and `examples` directory)

### CODE REVIEW

File: pkg/cfn/manager/api.go [hunk 1]
```diff
@@ -243,38 +245,55 @@ func (c *StackCollection) DeleteStack(name string, force bool) (*Stack, error) {
...
+// more then one error value on the channel, it's closed immediately after it is written to
+func (c *StackCollection) DeleteStackByNameSync(name string, errs chan error) error {
+	i, err := c.DeleteStackByName(name)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("waiting for stack %q to get deleted", *i.StackName)
+
+	go c.waitUntilStackIsDeleted(i, errs)
```
Comment (martina-if) [inline pkg/cfn/manager/api.go]: If we launch this in a goroutine, doesn't that mean this function will return immediately instead of waiting?

Comment (martina-if) [reply pkg/cfn/manager/api.go]: Same for the methods DeleteStackbySpecSync

File: pkg/cfn/manager/api.go [hunk 2]
```diff
@@ -243,38 +245,55 @@ func (c *StackCollection) DeleteStack(name string, force bool) (*Stack, error) {
...
+
+	return nil
+}
+
+// DeleteStackBySpec sends a request to delete the stack
+func (c *StackCollection) DeleteStackBySpec(s *Stack) (*Stack, error) {
```
Comment (martina-if) [inline pkg/cfn/manager/api.go]: Why does this return a stack?

File: pkg/cfn/manager/create_tasks.go [hunk 3]
```diff
@@ -6,9 +6,9 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-// CreateTasksForClusterWithNodeGroups defines all tasks required to create a cluster along
+// NewTasksToCreateClusterWithNodeGroups defines all tasks required to create a cluster along
 // with some nodegroups; see CreateAllNodeGroups for how onlyNodeGroupSubset works
-func (c *StackCollection) CreateTasksForClusterWithNodeGroups(onlyNodeGroupSubset sets.String) *TaskTree {
+func (c *StackCollection) NewTasksToCreateClusterWithNodeGroups(onlyNodeGroupSubset sets.String) *TaskTree {
```
Comment (martina-if) [inline pkg/cfn/manager/create_tasks.go]: I think maybe we could  drop the `WithNodeGroups` part. Or do we have code to create a cluster that doesn't create the nodegroup? Or maybe someone would be tempted to create the cluster and then again the nodegroups?

File: pkg/cfn/manager/api.go (hunk 1 above)
Comment (errordeveloper) [reply pkg/cfn/manager/api.go]: Yes, it returns, but `errs` is used for synchronisation.

File: pkg/cfn/manager/api.go (hunk 2 above)
Comment (errordeveloper) [reply pkg/cfn/manager/api.go]: So that we can pass it to `c.waitUntilStackIsDeleted`.

**Comment by errordeveloper** [comment]
@rndstr this is fully functional now, please have a look!

**Comment by rndstr** [comment]
It's not always doing what I would expect it to do, maybe I'm misunderstanding what the command with and without `--only-missing` does when given a `--config-file`

My expectations:
- without `--only-missing`: only nodegroups in the config file will be deleted
- with `--only-missing`: all nodegroups _not_ in the config file are deleted 

```
$ eksctl get nodegroups --cluster roli-dev
CLUSTER                 NODEGROUP       CREATED                 MIN SIZE        MAX SIZE        DESIRED CAPACITY        INSTANCE TYPE   IMAGE ID
roli-dev       ng1-public      2019-04-02T00:38:22Z    1               1               1                       m5.large        ami-0c28139856aaf9c3b
```

## with `--only-missing`

:x: Applying same config again (expected: no deletion)

`roli-dev.yaml`:
```
apiVersion: eksctl.io/v1alpha4
kind: ClusterConfig
metadata:
  name: roli-dev
  region: us-west-2
nodeGroups:
  - name: ng1-public
    instanceType: m5.large
    desiredCapacity: 1
```

Output:
```
$ ./eksctl delete nodegroup --only-missing --dry-run=true -f ./roli-dev.yaml
[ℹ]  1 nodegroup (ng1-public) was included
[ℹ]  (dry-run) 1 task: { delete nodegroup "ng1-public" [async] }
```

---
:x: Adding a new one (expected: no deletion)

`roli-dev.yaml`:
```
apiVersion: eksctl.io/v1alpha4
kind: ClusterConfig
metadata:
  name: roli-dev
  region: us-west-2
nodeGroups:
  - name: ng1-public
    instanceType: m5.large
    desiredCapacity: 1
  - name: ng1-public-new
    instanceType: m5.large
    desiredCapacity: 1
```

Output:
```
./eksctl delete nodegroup --only-missing --dry-run=true -f ./roli-dev.yaml
[ℹ]  2 nodegroups (ng1-public, ng1-public-new) were included
[ℹ]  (dry-run) 1 task: { delete nodegroup "ng1-public" [async] }
```

---
:heavy_check_mark: Renaming the nodegroup (expected: delete `ng1-public` nodegroup)

`roli-dev.yaml`:
```
apiVersion: eksctl.io/v1alpha4
kind: ClusterConfig
metadata:
  name: roli-fidelity-dev
  region: us-west-2
nodeGroups:
  - name: ng1-public-rename
    instanceType: m5.large
    desiredCapacity: 1
```

Output:
```
$ ./eksctl delete nodegroup --only-missing --dry-run=true -f ./roli-fidelity-dev.yaml
[ℹ]  nodegroup "ng1-public" present in the cluster, but missing from config file "./roli-fidelity-dev.yaml"
[ℹ]  1 nodegroup (ng1-public-rename) was excluded
[ℹ]  1 nodegroup (ng1-public) was included
[ℹ]  (dry-run) 1 task: { delete nodegroup "ng1-public" [async] }
```


## without `--only-missing`

:heavy_check_mark: Applying same config again (expected: delete `ng1-public`)

`roli-dev.yaml`:
```
apiVersion: eksctl.io/v1alpha4
kind: ClusterConfig
metadata:
  name: roli-dev
  region: us-west-2
nodeGroups:
  - name: ng1-public
    instanceType: m5.large
    desiredCapacity: 1
```

Output:
```
$ ./eksctl delete nodegroup --dry-run=true -f ./roli-fidelity-dev.yaml
[ℹ]  1 nodegroup (ng1-public) was included
[ℹ]  (dry-run) 1 task: { delete nodegroup "ng1-public" [async] }
```
---
:heavy_check_mark: Adding a new one (expected: delete `ng1-public`)

`roli-dev.yaml`:
```
apiVersion: eksctl.io/v1alpha4
kind: ClusterConfig
metadata:
  name: roli-dev
  region: us-west-2
nodeGroups:
  - name: ng1-public
    instanceType: m5.large
    desiredCapacity: 1
  - name: ng1-public-new
    instanceType: m5.large
    desiredCapacity: 1
```

Output:
```
$ ./eksctl delete nodegroup --dry-run=true -f ./roli-dev.yaml
[ℹ]  2 nodegroups (ng1-public, ng1-public-new) were included
[ℹ]  (dry-run) 1 task: { delete nodegroup "ng1-public" [async] }
```

---
:x: Renaming the nodegroup (expected: no deletion)

`roli-dev.yaml`:
```
apiVersion: eksctl.io/v1alpha4
kind: ClusterConfig
metadata:
  name: roli-fidelity-dev
  region: us-west-2
nodeGroups:
  - name: ng1-public-rename
    instanceType: m5.large
    desiredCapacity: 1
```

Output:
```
$ ./eksctl delete nodegroup --dry-run=true -f ./roli-fidelity-dev.yaml
[ℹ]  nodegroup "ng1-public" present in the cluster, but missing from config file "./roli-dev.yaml"
[ℹ]  2 nodegroups (ng1-public, ng1-public-rename) were included
[ℹ]  (dry-run) 1 task: { delete nodegroup "ng1-public" [async] }
```

**Comment by errordeveloper** [comment]
@rndstr I rewrote the logic that handle `--only-missing`, it should fix what you pointed out and another bug that I noticed when `--only-missing`+`--exclude` are used togethere.

**Comment by rndstr** [comment]
@errordeveloper works great now :ok_hand:

**Pull request merged**